### PR TITLE
Add document check job queue, abstract out lsp-editor%

### LINKED ISSUE #109: 100% CPU and never terminate for macro with infinite recursion
When I open a file with the following code in vscode:

```racket
#lang racket

(define-syntax (f stx)
  stx)

f
```

It reachs 100% CPU and doesn't terminate. I've been running it for a few minutes and it's still running.

I had this problem for a while, but just found a minimal sample to reproduce it.

### PR DESCRIPTION
**Pull request by 6cdh**
I did
* remove all `match` on struct `Doc`, Use struct accessor instead. So `doc` can contain extra hidden fields.
* add a extra middle layer `doc.rkt` between check-syntax, doc-trace and text-document, and move some code from `text-document.rkt` to `doc.rkt`. The `Doc-text`, `Doc-trace` should not be called in `text-document.rkt`, it should call a proxy method in `doc.rkt` instead.
* isolate `check-syntax` so that it's not called here and there. It's only called in `doc-run-check-syntax` now.
* clear `check-syntax`. It no longer sends diagnostics directly, instead, it returns diagnostics.
* use a unified editor class `lsp-editor%` instead of using `text%` or `racket:text%` everywhere. It would be useful if we write our own implementation to solve the gui problem.
* add a 90 seconds timeout for `expand`, should fix #109.
* improve the `queue-only-latest.rkt` implementation.
  * in the old design, there is at most a running `check-syntax` job and a pending job in the whole program. Any new incoming job will replace the pending job.
  * in the new design in `scheduler.rkt`, every document that is identified by uri can have a running job. Any new incoming job for this document will kill the old running job and run itself.
* make the initial trace object to be an empty `build-trace%` object not `#f`. It will fix many crash problem if `check-syntax` never succeed.
* other small improvements

The following issues will be fixed
* #101
* #97

### CODE REVIEW

File: check-syntax.rkt [hunk 1]
```diff
@@ -21,6 +20,13 @@
   (cond
     ;; typed racket support: don't report error summaries
     [(string-prefix? msg "Type Checker: Summary") (list)]
+    [(exn:fail:resource? exn)
+     (list (Diagnostic #:range (Range #:start (Pos #:line 0 #:char 0)
+                                      #:end (Pos #:line 0 #:char 0))
+                       #:severity Diag-Hint
+                       #:source "Expander"
+                       #:message "the expand time has exceeded the 90s limit.\
+                        Check if your macro is infinitely recursive."))]
```
Comment (dannypsnl) [inline check-syntax.rkt:29]: "is infinitely expanding"?

File: doc.rkt [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+;; the only place where really run check-syntax
+(define (doc-run-check-syntax doc)
+  (define (task)
+    (set-Doc-checked?! doc #f)
+    (match-define (list new-trace diags)
+      (check-syntax (uri->path (Doc-uri doc)) (Doc-text doc) (Doc-trace doc)))
+    (send-diagnostics doc diags)
+    (set-Doc-trace! doc new-trace)
+
+    (set-Doc-checked?! doc #t))
```
Comment (dannypsnl) [inline doc.rkt:41]: Could this function be called in concurrency? If so, then the `checked?` modification is not safe, need a lock here

File: doc.rkt [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+(define (doc-line-start-pos doc line)
+  (send (Doc-text doc) line-start-pos line))
+
+(define (doc-line-end-pos doc line)
+  (send (Doc-text doc) line-end-pos line))
+
+(define (doc-endpos doc)
+  (send (Doc-text doc) end-pos))
+
+(define (doc-find-containing-paren doc pos)
```
Comment (dannypsnl) [inline doc.rkt:109]: What's the purpose of this function? Checking paren balancing and stop invalid file?

File: scheduler.rkt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+;; no matter if the old one is running or completed
+(define (schedule)
+  (define open-doc (make-hash))
+  (let loop ()
+    (sync (handle-evt incoming-jobs-ch
+                      (λ (data)
+                        (match-define (cons path task) data)
+                        (when (hash-has-key? open-doc path)
+                          (define th (hash-ref open-doc path))
+                          (when (not (thread-dead? th))
```
Comment (dannypsnl) [inline scheduler.rkt:18]: Maybe use `unless (thread-dead? th)`?

File: text-document.rkt [hunk 5]
```diff
@@ -118,12 +121,14 @@
...
-     (unless (uri-is-path? uri)
-       (error 'hover "uri is not a path"))
-     (match-define (doc doc-text doc-trace)
-       (hash-ref open-docs (string->symbol uri)))
+     #:when (uri-is-path? uri)
+
+     (define this-doc (hash-ref open-docs (string->symbol uri)))
+
+     (define doc-trace (Doc-trace this-doc))
+
```
Comment (dannypsnl) [inline text-document.rkt:129]: Misc. I think we don't need these blank line?

File: doc.rkt (hunk 2 above)
Comment (6cdh) [reply doc.rkt:41]: No. There's only one `task` running for a given document at anytime.
The `set-Doc-*` methods are expected to be safe.

File: doc.rkt (hunk 3 above)
Comment (6cdh) [reply doc.rkt:109]: It's a new wrapper for the existing `find-containing-paren`. The function finds the closest pair of brackets the containing `pos` If I understand it correctly.

**Review by dannypsnl — APPROVED** [review]

**Comment by dannypsnl** [comment]
I think the title can be "Add document check job queue, abstract out `lsp-editor%`"

File: editor.rkt [hunk 6]
```diff
@@ -9,16 +9,32 @@
...
 
     ;; replace text at (range start end) with str
     (define/public (replace str start end)
-      (send core insert str start end))
+      (with-handlers ([exn?
+                       (λ _
+                         (reload!)
+                         ;; only retry once
+                         (send core insert str start end))])
+        (send core insert str start end)))
```
Comment (dannypsnl) [inline editor.rkt:37]: `send core replace ...`?

**Review by dannypsnl — CHANGES_REQUESTED** [review]

**Comment by 6cdh** [comment]
The new `fix editor crash` commit fixed a random editor crash with this message:
```
sequence-contract-violation: negative: method insert cannot be called
```

It's a well known bug that no one knows how it happened. https://github.com/racket/drracket/issues/522
I found it today and added a workaround that recreate a new `text%` to replace to old one.

**Comment by 6cdh** [comment]
Everything should work fine right now to be merged.

File: editor.rkt (hunk 6 above)
Comment (6cdh) [reply editor.rkt:37]: There is no such a method for `text%`.

Comment (dannypsnl) [reply editor.rkt:37]: Oh, I get the difference.

**Review by dannypsnl — APPROVED** [review]

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**